## Import Required Libraries

In [436]:
import numpy as np
import pandas as pd

## Import the file to Jupyter

In [438]:
df = pd.read_csv("covid_19_data.csv")

In [439]:
df.head(10)


SNo ObservationDate Province/State  Country/Region      Last Update  \
0    1       1/22/2020          Anhui  Mainland China  1/22/2020 17:00   
1    2       1/22/2020        Beijing  Mainland China  1/22/2020 17:00   
2    3       1/22/2020      Chongqing  Mainland China  1/22/2020 17:00   
3    4       1/22/2020         Fujian  Mainland China  1/22/2020 17:00   
4    5       1/22/2020          Gansu  Mainland China  1/22/2020 17:00   
5    6       1/22/2020      Guangdong  Mainland China  1/22/2020 17:00   
6    7       1/22/2020        Guangxi  Mainland China  1/22/2020 17:00   
7    8       1/22/2020        Guizhou  Mainland China  1/22/2020 17:00   
8    9       1/22/2020         Hainan  Mainland China  1/22/2020 17:00   
9   10       1/22/2020          Hebei  Mainland China  1/22/2020 17:00   

   Confirmed  Deaths  Recovered  Unnamed: 8  Unnamed: 9  Unnamed: 10  
0          1       0          0         NaN         NaN          NaN  
1         14       0          0         NaN         NaN          NaN  
2          6       0          0         NaN         NaN          NaN  
3          1       0          0         NaN         NaN          NaN  
4          0       0          0         NaN         NaN          NaN  
5         26       0          0         NaN         NaN          NaN  
6          2       0          0         NaN         NaN          NaN  
7          1       0          0         NaN         NaN          NaN  
8          4       0          0         NaN         NaN          NaN  
9          1       0          0         NaN         NaN          NaN

## Drop unwanted columns

In [440]:
df1 = df.drop(["Unnamed: 8","Unnamed: 9","Unnamed: 10"], axis = 1)
df2 = df1.set_index('SNo')

In [441]:
df2.head()

ObservationDate Province/State  Country/Region      Last Update  \
SNo                                                                   
1         1/22/2020          Anhui  Mainland China  1/22/2020 17:00   
2         1/22/2020        Beijing  Mainland China  1/22/2020 17:00   
3         1/22/2020      Chongqing  Mainland China  1/22/2020 17:00   
4         1/22/2020         Fujian  Mainland China  1/22/2020 17:00   
5         1/22/2020          Gansu  Mainland China  1/22/2020 17:00   

     Confirmed  Deaths  Recovered  
SNo                                
1            1       0          0  
2           14       0          0  
3            6       0          0  
4            1       0          0  
5            0       0          0

## Check the Data- Types

In [443]:
dataTypeSeries = df2.dtypes
dataTypeSeries

ObservationDate    object
Province/State     object
Country/Region     object
Last Update        object
Confirmed           int64
Deaths              int64
Recovered           int64
dtype: object

## Convert the date-time information - Last Update and ObservationDate

In [451]:
df2["Last Update"]

SNo
1            1/22/2020 17:00
2            1/22/2020 17:00
3            1/22/2020 17:00
4            1/22/2020 17:00
5            1/22/2020 17:00
6            1/22/2020 17:00
7            1/22/2020 17:00
8            1/22/2020 17:00
9            1/22/2020 17:00
10           1/22/2020 17:00
11           1/22/2020 17:00
12           1/22/2020 17:00
13           1/22/2020 17:00
14           1/22/2020 17:00
15           1/22/2020 17:00
16           1/22/2020 17:00
17           1/22/2020 17:00
18           1/22/2020 17:00
19           1/22/2020 17:00
20           1/22/2020 17:00
21           1/22/2020 17:00
22           1/22/2020 17:00
23           1/22/2020 17:00
24           1/22/2020 17:00
25           1/22/2020 17:00
26           1/22/2020 17:00
27           1/22/2020 17:00
28           1/22/2020 17:00
29           1/22/2020 17:00
30           1/22/2020 17:00
31           1/22/2020 17:00
32           1/22/2020 17:00
33           1/22/2020 17:00
34           1/22/2020 17:00
35        

In [452]:
type(df2["Last Update"])

pandas.core.series.Series

In [453]:
type(df2["Last Update"][1])

str

In [454]:
#df2["Date"] = pd.to_datetime(df3["Last Update"].date())
#df2["date"]=pd.to_datetime(df3["Last Update"]).date()
#pd.to_datetime(df2['Last Update'], utc=False)

df2['Last_Update'] = pd.to_datetime(df2['Last Update']).dt.date


In [455]:

df2['ObservedDate'] = pd.to_datetime(df2['ObservationDate']).dt.normalize()

In [456]:
df2.head()

ObservationDate Province/State  Country/Region      Last Update  \
SNo                                                                   
1         1/22/2020          Anhui  Mainland China  1/22/2020 17:00   
2         1/22/2020        Beijing  Mainland China  1/22/2020 17:00   
3         1/22/2020      Chongqing  Mainland China  1/22/2020 17:00   
4         1/22/2020         Fujian  Mainland China  1/22/2020 17:00   
5         1/22/2020          Gansu  Mainland China  1/22/2020 17:00   

     Confirmed  Deaths  Recovered Last_Update ObservedDate  
SNo                                                         
1            1       0          0  2020-01-22   2020-01-22  
2           14       0          0  2020-01-22   2020-01-22  
3            6       0          0  2020-01-22   2020-01-22  
4            1       0          0  2020-01-22   2020-01-22  
5            0       0          0  2020-01-22   2020-01-22

## Drop the old variables - Last Update and ObservationDate

In [457]:
df3 = df2.drop(["Last Update","ObservationDate"], axis = 1 )

## Renaming and arranging the variables 

In [458]:
df3.columns.values

array(['Province/State', 'Country/Region', 'Confirmed', 'Deaths',
       'Recovered', 'Last_Update', 'ObservedDate'], dtype=object)

In [459]:
#df2["Last_Update"] = df2["Last Update"]
df3['Province_State'] = df3['Province/State']
df3['Country_Region'] = df3['Country/Region']

df4 = ['ObservedDate', 'Province_State', 'Country_Region',
       'Last_Update', 'Confirmed', 'Deaths', 'Recovered']

In [460]:
df2 = df3[df4]

In [461]:
df2.head()

ObservedDate Province_State  Country_Region Last_Update  Confirmed  \
SNo                                                                      
1     2020-01-22          Anhui  Mainland China  2020-01-22          1   
2     2020-01-22        Beijing  Mainland China  2020-01-22         14   
3     2020-01-22      Chongqing  Mainland China  2020-01-22          6   
4     2020-01-22         Fujian  Mainland China  2020-01-22          1   
5     2020-01-22          Gansu  Mainland China  2020-01-22          0   

     Deaths  Recovered  
SNo                     
1         0          0  
2         0          0  
3         0          0  
4         0          0  
5         0          0

## Describing the dataframe

describe function by default will only gives the describtion of the numerical variables. Therefore we have to add <b>" Include = all" </b>

In [462]:
df2.describe()

Confirmed        Deaths      Recovered
count   48095.000000  48095.000000   48095.000000
mean     8032.769456    488.668510    3169.632041
std     27135.172613   2414.064891   18424.263956
min         0.000000      0.000000       0.000000
25%        63.000000      1.000000       0.000000
50%       593.000000      9.000000      55.000000
75%      3432.000000    105.000000     789.000000
max    405843.000000  41128.000000  640198.000000

In [463]:
df2.describe(include = "all")

ObservedDate                Province_State Country_Region  \
count                 48095                         28390          48095   
unique                  153                           709            223   
top     2020-06-16 00:00:00  Diamond Princess cruise ship             US   
freq                    729                           242           6962   
first   2020-01-22 00:00:00                           NaN            NaN   
last    2020-06-22 00:00:00                           NaN            NaN   
mean                    NaN                           NaN            NaN   
std                     NaN                           NaN            NaN   
min                     NaN                           NaN            NaN   
25%                     NaN                           NaN            NaN   
50%                     NaN                           NaN            NaN   
75%                     NaN                           NaN            NaN   
max                     NaN                           NaN            NaN   

       Last_Update      Confirmed        Deaths      Recovered  
count        48095   48095.000000  48095.000000   48095.000000  
unique         149            NaN           NaN            NaN  
top     2020-03-08            NaN           NaN            NaN  
freq          1481            NaN           NaN            NaN  
first          NaN            NaN           NaN            NaN  
last           NaN            NaN           NaN            NaN  
mean           NaN    8032.769456    488.668510    3169.632041  
std            NaN   27135.172613   2414.064891   18424.263956  
min            NaN       0.000000      0.000000       0.000000  
25%            NaN      63.000000      1.000000       0.000000  
50%            NaN     593.000000      9.000000      55.000000  
75%            NaN    3432.000000    105.000000     789.000000  
max            NaN  405843.000000  41128.000000  640198.000000

## Checking for null values

Province_State has 19705 missing values. However, we can ignore the variable as it's useless for our current analysis

In [466]:
df2.isnull().sum()

ObservedDate          0
Province_State    19705
Country_Region        0
Last_Update           0
Confirmed             0
Deaths                0
Recovered             0
dtype: int64

#### 1.  Calculate the average death cases of confirmed, deaths, recovered

<ul> 

<li> <b> Confirmed </b> average : 8032.769456 , mode : 593.000000</li>
<li> <b> Confirmed </b>average : 488.668510 , mode : 9.000000 </li>
<li> <b> Confirmed </b>average : 3169.632041 , mode : 55.000000</li>
</ul>

####  2. What are the maximum cases confirmed in a single day

In [467]:
df2.Confirmed.max()

405843

#### 3. Which Region contains maximum cases Confirmed on that single day?

In [468]:
df3 = df2.groupby(['Country_Region'])['Confirmed'].max()
pd.set_option('display.max_rows', None)


In [469]:
df3.sort_values(inplace = False)

Country_Region
 Azerbaijan                              1
Cayman Islands                           1
Cape Verde                               1
East Timor                               1
Gambia, The                              1
Gibraltar                                1
Greenland                                1
Guernsey                                 1
Curacao                                  1
North Ireland                            1
Republic of the Congo                    1
Channel Islands                          1
Vatican City                             1
The Gambia                               1
The Bahamas                              1
('St. Martin',)                          2
Faroe Islands                            2
St. Martin                               2
Jersey                                   2
Saint Barthelemy                         3
Puerto Rico                              3
Guam                                     3
Aruba                                  

#### Answer : 

Russia has maximum cases <strong>confirmed</strong> with the total <strong> 405843</strong> cases registed in a day

#### 4. On which Date contains maximum cases Confirmed on that single day

In [470]:
df4 = df2.groupby(['Last_Update'])['Confirmed'].max()
pd.set_option('display.max_rows', None)
df4

Last_Update
2020-01-22       444
2020-01-23       444
2020-01-24       549
2020-01-25       761
2020-01-26      1058
2020-01-27      1423
2020-01-28      3554
2020-01-29      3554
2020-01-30      4903
2020-01-31      5806
2020-02-01      7153
2020-02-02     11177
2020-02-03     13522
2020-02-04     16678
2020-02-05     19665
2020-02-06     22112
2020-02-07     24953
2020-02-08     27100
2020-02-09     29631
2020-02-10     31728
2020-02-11     33366
2020-02-12     33366
2020-02-13     48206
2020-02-14     54406
2020-02-15     56249
2020-02-16     58182
2020-02-17     59989
2020-02-18     61682
2020-02-19     62031
2020-02-20     62442
2020-02-21     62662
2020-02-22     64084
2020-02-23     64084
2020-02-24     64287
2020-02-25     64786
2020-02-26     65187
2020-02-27     65596
2020-02-28     65914
2020-02-29     66337
2020-03-01     66907
2020-03-02     67103
2020-03-03     67217
2020-03-04     67332
2020-03-05     67466
2020-03-06     67592
2020-03-07     67666
2020-03-08    101739
2

In [471]:
df4.sort_values(inplace = False)

Last_Update
2020-01-22       444
2020-01-23       444
2020-01-24       549
2020-01-25       761
2020-01-26      1058
2020-01-27      1423
2020-01-28      3554
2020-01-29      3554
2020-01-30      4903
2020-01-31      5806
2020-02-01      7153
2020-02-02     11177
2020-02-03     13522
2020-02-04     16678
2020-02-05     19665
2020-02-06     22112
2020-02-07     24953
2020-02-08     27100
2020-02-09     29631
2020-02-10     31728
2020-02-11     33366
2020-02-12     33366
2020-02-13     48206
2020-02-14     54406
2020-02-15     56249
2020-02-16     58182
2020-02-17     59989
2020-02-18     61682
2020-02-19     62031
2020-02-20     62442
2020-02-21     62662
2020-02-22     64084
2020-02-23     64084
2020-02-24     64287
2020-02-25     64786
2020-02-26     65187
2020-02-27     65596
2020-02-28     65914
2020-02-29     66337
2020-03-01     66907
2020-03-02     67103
2020-03-03     67217
2020-03-04     67332
2020-03-05     67466
2020-03-06     67592
2020-03-07     67666
2020-03-09     67743
2

#### Answer : 

 Total <strong>405843</strong> cases has  only been confirmed on <strong>1st July 2020</strong> 

#### 5. Construct a table using Region variable with groupby

In [472]:
df2.groupby(["Country_Region"])['Confirmed','Deaths','Recovered'].sum()

<ipython-input-472-2b55be8a8c3f>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df2.groupby(["Country_Region"])['Confirmed','Deaths','Recovered'].sum()


Confirmed   Deaths  Recovered
Country_Region                                                 
 Azerbaijan                               1        0          0
('St. Martin',)                           2        0          0
Afghanistan                          760916    15272     120898
Albania                               79262     2595      52058
Algeria                              520365    42923     294273
Andorra                               64062     3603      41031
Angola                                 5052      276       1600
Antigua and Barbuda                    2026      221       1139
Argentina                           1001549    32423     297560
Armenia                              525341     8008     200510
Aruba                                    19        0          0
Australia                            602746     7324     434132
Austria                             1377553    46666    1042563
Azerbaijan                           340330     4114     193876
Bahamas                                6865      818       2932
Bahamas, The                             10        0          0
Bahrain                              642977     1409     384054
Bangladesh                          2580809    36670     655206
Barbados                               7107      493       4280
Belarus                             2179240    12643     932612
Belgium                             4083618   621512    1027215
Belize                                 1431      152        902
Benin                                 17088      238       7335
Bhutan                                 2088        0        649
Bolivia                              528229    18727      79792
Bosnia and Herzegovina               168570     8652      96733
Botswana                               2547       84       1004
Brazil                             26892505  1479988   12723188
Brunei                                13233      121      10432
Bulgaria                             164512     8236      59408
Burkina Faso                          60742     3728      44454
Burma                                 13897      429       6675
Burundi                                3037       71       1650
Cabo Verde                            25641      254       9434
Cambodia                              11438        0       9245
Cameroon                             335737    10184     176153
Canada                              5526702   392407    2732005
Cape Verde                                1        0          0
Cayman Islands                            3        0          0
Central African Republic              55639      230       5416
Chad                                  33399     2917      20372
Channel Islands                           1        0          0
Chile                               5461026    79249    3528713
Colombia                            1622435    57053     526928
Comoros                                4885       92       2337
Congo (Brazzaville)                   32385     1120       9195
Congo (Kinshasa)                     157329     4527      21217
Costa Rica                            82629      654      37597
Croatia                              173940     6278     121642
Cuba                                 132211     5122      91826
Curacao                                   2        0          0
Cyprus                                71563     1297      35891
Czech Republic                       690858    20957     355717
Denmark                              845495    39310     632414
Diamond Princess                      64080     1120      57613
Djibouti                             165686     1079      92286
Dominica                               1429        0       1029
Dominican Republic                   978587    31458     457795
East Timor                                1        0          0
Ecuador                             2364465   173524     724976
Egypt                               1395619    59371     363489
El

#### 6. From the overall data, which region contains maximum cases Confirmed?

In [473]:
df5 = df2.groupby(['Country_Region'])['Confirmed'].sum()
pd.set_option('display.max_rows', None)
df5.sort_values(inplace = False)

Country_Region
 Azerbaijan                                 1
Channel Islands                             1
East Timor                                  1
North Ireland                               1
Republic of the Congo                       1
Cape Verde                                  1
The Gambia                                  1
('St. Martin',)                             2
Curacao                                     2
St. Martin                                  2
The Bahamas                                 3
Puerto Rico                                 3
Cayman Islands                              3
Guernsey                                    3
Greenland                                   3
Gambia, The                                 4
Vatican City                                4
Jersey                                      6
Guam                                        6
Gibraltar                                   7
Bahamas, The                               10
Faroe Islands      

 #### Answer: 
 
 The <strong>US</strong> has registered maximum cases. Total cases are: <strong>110818653</strong>

#### 7. Create a subset of only India from the Region column

In [474]:
df6 = df2[df2['Country_Region'] == "India"]
df6.head()

ObservedDate Province_State Country_Region Last_Update  Confirmed  Deaths  \
SNo                                                                             
431   2020-01-30            NaN          India  2020-01-30          1       0   
492   2020-01-31            NaN          India  2020-01-31          1       0   
548   2020-02-01            NaN          India  2020-01-31          1       0   
608   2020-02-02            NaN          India  2020-02-02          2       0   
673   2020-02-03            NaN          India  2020-02-03          3       0   

     Recovered  
SNo             
431          0  
492          0  
548          0  
608          0  
673          0

#### 8. Total number of cases confirmed in India

In [475]:
df6['Confirmed'].sum()

10295111

#### 9. On which day India confirmed the maximum number of cases registered

In [476]:
df7 = df6.groupby(['Last_Update'])['Confirmed'].max()
df7.sort_values( inplace = False)

Last_Update
2020-01-30         1
2020-01-31         1
2020-02-02         2
2020-02-03         3
2020-02-13         3
2020-02-16         3
2020-03-02         5
2020-03-04        28
2020-03-05        30
2020-03-06        31
2020-03-07        34
2020-03-09        43
2020-03-10        56
2020-03-12        73
2020-03-11        82
2020-03-14       102
2020-03-15       113
2020-03-16       119
2020-03-17       142
2020-03-18       156
2020-03-19       194
2020-03-20       244
2020-03-21       330
2020-03-23       499
2020-03-24       536
2020-03-25       657
2020-03-26       727
2020-03-27       887
2020-03-08      1251
2020-03-31      1397
2020-04-01      1998
2020-04-02      2543
2020-04-03      2567
2020-04-04      3082
2020-04-05      3588
2020-04-06      4778
2020-04-07      5311
2020-04-08      5916
2020-04-09      6725
2020-04-10      7598
2020-04-11      8446
2020-04-12      9205
2020-04-13     10453
2020-04-14     11487
2020-04-15     12322
2020-04-16     13430
2020-04-17     14352
2

#### Answer: 

India confirmed maximum cases on <strong>10th July 2020</strong> with total <strong>276146</strong> cases

#### 10. Calculate the percentages of death cases from total confirmed cases in India

In [477]:
df6['Deaths'].mean()

520.0916666666667